File: go/6-functions.go
package main
import "fmt"

func dodaj(a, b int ) int {
	return a+b
}

func echo(e int) {
	fmt.Printf("echo e=%d\n", e); 
	e--
	if (e > 0) { echo(e) }
}
type operation func(int, int ) int

func increment(what *int, to *int) {
    fmt.Printf("increment: received *what=%d what=%p *to=%d to=%p\n", *what, what, *to, to)
    (*what)++
    *to++
    fmt.Printf("increment: returning *what=%d what=%p *to=%d, to=%p\n", *what, what, *to, to )
}

func main() {
	a:=2
	b:=3
fmt.Printf("%d + %d = %d\n", a, b, add(a, b))

echo(10)
var fun_ptr operation
fun_ptr=add
fmt.Printf("result of operation on %d and %d=%d\n", a, b, fun_ptr(a,b))

fmt.Printf("a = %d b=%d\n", a, b)
increment(&a, &b)
fmt.Printf("a after increment= %d\n", a)
fmt.Printf("b after increment= %d\n", b)
}

File: lua/6-functions.lua
#!/bin/lua
function add(a, b)
    return a+b
end

function echo(e)
    print("echo:", e)
    e=e-1
    if e > 0
    then 
        echo(e) 
    end
end

function increment(what, to)
    print("increment: received what=", what, "to=", to)
    what=what+1
    to=to+1
    print("increment: returning what=", what, "to=", to)
    return what, to
end

a=2
b=3
print(a, "+", b, "=", add(a, b))

echo(10)

operation=add;
print("result of operation on", a, "and", b, "=", operation(a,b))

operation1 = function(a, b)
    return a-b;
end

print("operation1 subtract on", a, "and", b,"=", operation1(a,b))

print("a=", a, "b=", b)
a, b = increment(a, b)
print("a after increment=", a)
print("b after increment=", b)

foo=function(f)
    print("function foo: f=", f)
    if (type(f) == "table") 
    then
        print("f is a table")
    else
        print("f is a ", type(f))
    end
end

foo(123)
foo {456}
foo({789})
print("foo type=", type(foo))


